gossip/api/types.go

```diff
@@ -0,0 +1,231 @@
+// RFC 5246 notation:
+//   struct {
+//      opaque source_id<1,65535>;
+//      opaque head_data<1,65535>;
+//      opaque source_signature<0,65535>;
+//      uint64 hub_timestamp;
+//   } TimestampedEntry;
+type TimestampedEntry struct {
+	SourceID        []byte `tls:"minlen:1,maxlen:65535"`
+	BlobData        []byte `tls:"minlen:1,maxlen:65535"`
+	SourceSignature []byte `tls:"minlen:0,maxlen:65535"`
+	HubTimestamp    uint64 // in nanoseconds since UNIX epoch
+}
+
+// SignedGossipTimestamp holds a timestamped entry in the gossip hub together with a
+// signature from the hub.  RFC 5246 notation:
+//   struct {
+//      TimestampedEntry timestamped_entry;
+//      opaque hub_signature<1,65535>;
+//   } SignedGossipTimestamp;
+type SignedGossipTimestamp struct {
+	TimestampedEntry TimestampedEntry
+	// HubSignature is over the TLS-encoding of GossipTimestamp.
+	HubSignature []byte `tls:"minlen:1,maxlen:65535"`
+}
+
+// Retrieve Latest Signed Tree Head for Hub
+// GET https://<hub server>/gossip/v0/get-sth
+// Inputs: None
+// Outputs:
+//    head_data: TLS-encoded HubTreeHead.
+//    hub_signature: signature over head_data.
+
+// GetSTHPath is the final path component for this entrypoint.
+const GetSTHPath = "get-sth"
+
+// GetSTHResponse represents the JSON response to the get-sth GET method.
+type GetSTHResponse struct {
+	// HeadData holds the TLS-encoding of a HubTreeHead.
+	TreeHeadData []byte `json:"head_data"`
+	// HubSignature is a signature over TreeHeadData from the hub.
+	HubSignature []byte `json:"hub_signature"`
+}
+
+// HubTreeHead describes a head of the Hub's Merkle tree. RFC 5246 notation:
+//   struct {
+//     uint64 tree_size;
+//     uint64 timestamp;
+//     opaque root_hash<1,65535>;
+//   } HubTreeHead;
+// TODO(drysdale): shift this to being a versioned structure to allow for
+// future expansion (cf. github.com/google/trillian/types.LogRoot).
+type HubTreeHead struct {
+	TreeSize  uint64
+	Timestamp uint64 // in nanoseconds since UNIX epoch
+	RootHash  []byte `tls:"minlen:1,maxlen:65535"`
+}
+
+// SignedHubTreeHead describes a head of the Hub's Merkle tree, together with
+// a signature from the hub.
+type SignedHubTreeHead struct {
+	TreeHead     HubTreeHead
+	HubSignature []byte // over tls.Marshal(TreeHead)
+}
+
+// Retrieve Consistency Proof between Tree Heads
+// GET https://<hub server>/gossip/v0/get-sth-consistency
+// Inputs:
+//    first:  The tree_size of the first tree, in decimal.
+//    second:  The tree_size of the second tree, in decimal.
+// Both tree sizes must be from existing v1 STHs (Signed Tree Heads).
+// Outputs:
+//    consistency:  An array of Merkle Tree nodes, base64 encoded.
+
+// GetSTHConsistencyPath is the final path component for this entrypoint.
+const GetSTHConsistencyPath = "get-sth-consistency"
+const (
+	// GetSTHConsistencyFirst is the first parameter name.
+	GetSTHConsistencyFirst = "first"
```

Comment:
For consistency with proto3 and better typing , should these fields be inside a `GetSTHConsistencyRequest` struct?

Reply:
Such a struct would currently be unused (there's no way to encode a `struct` to a GET request), so I don't think it makes sense to add it at the moment.

Reply:
Well, this whole thing is in the -examples repo and marked as v0, so I think we can re-jig it however we like later.  I'll make sure there's some warnings about that in the README too.